### ISSUE TITLE
Overriding pre-defined spaces: TypeError: stoploss_space() missing 1 required positional argument: 'self'

### ISSUE THREAD

**Issue by EnzovdWetering** [freqtrade/freqtrade#5543]
## Describe your environment

  * Operating system: Ubuntu 20.04.3 LTS
  * Python Version: 3.8.10 (`python -V`)
  * CCXT version: 1.55.28 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2021.8 (`freqtrade -V` or `docker-compose run --rm freqtrade -V` for Freqtrade running in docker)
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:

I am trying to use a custom seach-space for stoploss to hyperopt it however, when I add the hyperopt class and everything below it to my Strategy, I get the error below:

```
class ElliotV8_original_ichiv3_roi(IStrategy):
    class HyperOpt:
        # Define a custom stoploss space.
        def stoploss_space(self):
            return [SKDecimal(-0.05, -0.01, decimals=3, name='stoploss')]
```

I got this template from the documentation found here: https://www.freqtrade.io/en/stable/advanced-hyperopt/#pverriding-pre-defined-spaces

```
2021-09-07 18:26:25,894 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/enzo/freqtrade/freqtrade/main.py", line 37, in main
    return_code = args['func'](args)
  File "/home/enzo/freqtrade/freqtrade/commands/optimize_commands.py", line 87, in start_hyperopt
    hyperopt.start()
  File "/home/enzo/freqtrade/freqtrade/optimize/hyperopt.py", line 426, in start
    self.init_spaces()
  File "/home/enzo/freqtrade/freqtrade/optimize/hyperopt.py", line 280, in init_spaces
    self.stoploss_space = self.custom_hyperopt.stoploss_space()
  File "/home/enzo/freqtrade/freqtrade/optimize/hyperopt_auto.py", line 86, in stoploss_space
    return self._get_func('stoploss_space')()
TypeError: stoploss_space() missing 1 required positional argument: 'self'
```

### Steps to reproduce:

  1. Add the HyperOpt class to ElliotV8_original_ichiv3_roi
  2. Run this command to start the hyperopt:  `freqtrade hyperopt --spaces stoploss --strategy ElliotV8_original_ichiv3_roi -e 100000 --config config-staticparis.json --hyperopt-loss SortinoHyperOptLossDaily --enable-protections --timerange 20210816-  --print-all`
  3. The stoploss_space() missing 1 required positional argument: 'self' error shows up. 
 
I also tried this on the develop: 880474594e5357eea9b47b8d1c8449a968bce4c3 but I observed the same results.
  
### Observed Results:

See above, also please note that my main goal is to use a custom seach space for ROI, but for throubleshoothing, I decided to use the template as found on the feqtrade documenation page. 

### Relevant code exceptions or logs

  ```
# --- Do not remove these libs ---
from freqtrade.strategy.interface import IStrategy
from typing import Dict, List
from functools import reduce
from pandas import DataFrame
# --------------------------------
import talib.abstract as ta
import numpy as np
import freqtrade.vendor.qtpylib.indicators as qtpylib
import datetime
from technical.util import resample_to_interval, resampled_merge
from datetime import datetime, timedelta
from freqtrade.persistence import Trade
from freqtrade.strategy import stoploss_from_open, merge_informative_pair, DecimalParameter, IntParameter, \
    CategoricalParameter
from freqtrade.optimize.space import Categorical, Dimension, Integer, SKDecimal, Real  # noqa
import technical.indicators as ftt

# @Rallipanos # changes by IcHiAT

# Buy hyperspace params:
buy_params = {
    "base_nb_candles_buy": 61,  
    "ewo_high": 3.554,  
    "ewo_low": -19.385,  
    "low_offset": 0.987,
    "rsi_buy": 38,  
}

# Sell hyperspace params:
sell_params = {
    "base_nb_candles_sell": 43, 
    "high_offset": 0.998, 
    "high_offset_2": 1.097,
}


def EWO(dataframe, ema_length=5, ema2_length=3):
    df = dataframe.copy()
    ema1 = ta.EMA(df, timeperiod=ema_length)
    ema2 = ta.EMA(df, timeperiod=ema2_length)
    emadif = (ema1 - ema2) / df['close'] * 100
    return emadif


class ElliotV8_original_ichiv3_roi(IStrategy):
    class HyperOpt:
        # Define a custom stoploss space.
        def stoploss_space(self):
            return [SKDecimal(-0.05, -0.01, decimals=3, name='stoploss')]

    INTERFACE_VERSION = 2
    """
    # ROI table:
    minimal_roi = {
        "0": 0.08,
        "20": 0.04,
        "40": 0.032,
        "87": 0.016,
        "201": 0,
        "202": -1
    }
    """

    @property
    def protections(self):
        return [
            {
                "method": "CooldownPeriod",
                "stop_duration_candles": 5
            },
            {
                "method": "MaxDrawdown",
                "lookback_period_candles": 48,
                "trade_limit": 20,
                "stop_duration_candles": 4,
                "max_allowed_drawdown": 0.2
            },
            {
                "method": "StoplossGuard",
                "lookback_period_candles": 24,
                "trade_limit": 4,
                "stop_duration_candles": 2,
                "only_per_pair": False
            },
            {
                "method": "LowProfitPairs",
                "lookback_period_candles": 6,
                "trade_limit": 2,
                "stop_duration_candles": 60,
                "required_profit": 0.02
            },
            {
                "method": "LowProfitPairs",
                "lookback_period_candles": 24,
                "trade_limit": 4,
                "stop_duration_candles": 2,
                "required_profit": 0.01
            }
        ]

    # ROI table:
    minimal_roi = {
        "0": 0.99,
        "200": -1
    }

    # Stoploss:
    stoploss = -0.291

    # SMAOffset
    base_nb_candles_buy = IntParameter(5, 80, default=buy_params['base_nb_candles_buy'], space='buy', optimize=True)
    base_nb_candles_sell = IntParameter(5, 80, default=sell_params['base_nb_candles_sell'], space='sell', optimize=True)
    low_offset = DecimalParameter(0.9, 0.99, default=buy_params['low_offset'], space='buy', optimize=True)
    high_offset = DecimalParameter(0.95, 1.1, default=sell_params['high_offset'], space='sell', optimize=True)
    high_offset_2 = DecimalParameter(0.99, 1.5, default=sell_params['high_offset_2'], space='sell', optimize=True)

    # Protection
    fast_ewo = 50
    slow_ewo = 200
    ewo_low = DecimalParameter(-20.0, -8.0, default=buy_params['ewo_low'], space='buy', optimize=True)
    ewo_high = DecimalParameter(2.0, 12.0, default=buy_params['ewo_high'], space='buy', optimize=True)
    rsi_buy = IntParameter(30, 70, default=buy_params['rsi_buy'], space='buy', optimize=True)

    # Trailing stop:
    trailing_stop = True
    trailing_stop_positive = 0.001
    trailing_stop_positive_offset = 0.02
    trailing_only_offset_is_reached = True

    # Sell signal
    use_sell_signal = True
    sell_profit_only = True
    sell_profit_offset = 0.01
    ignore_roi_if_buy_signal = False

    ## Optional order time in force.
    order_time_in_force = {
        'buy': 'gtc',
        'sell': 'gtc'
    }

    # Optimal timeframe for the strategy
    timeframe = '5m'
    inf_1h = '1h'

    process_only_new_candles = True
    startup_candle_count = 400

    plot_config = {
        'main_plot': {
            'ma_buy': {'color': 'orange'},
            'ma_sell': {'color': 'orange'},
        },
    }

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        # Calculate all ma_buy values
        for val in self.base_nb_candles_buy.range:
            dataframe[f'ma_buy_{val}'] = ta.EMA(dataframe, timeperiod=val)

        # Calculate all ma_sell values
        for val in self.base_nb_candles_sell.range:
            dataframe[f'ma_sell_{val}'] = ta.EMA(dataframe, timeperiod=val)

        dataframe['hma_50'] = qtpylib.hull_moving_average(dataframe['close'], window=50)

        dataframe['sma_9'] = ta.SMA(dataframe, timeperiod=9)
        # Elliot
        dataframe['EWO'] = EWO(dataframe, self.fast_ewo, self.slow_ewo)

        # RSI
        dataframe['rsi'] = ta.RSI(dataframe, timeperiod=14)
        dataframe['rsi_fast'] = ta.RSI(dataframe, timeperiod=4)
        dataframe['rsi_slow'] = ta.RSI(dataframe, timeperiod=20)

        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        conditions = []

        conditions.append(
            (

                    (dataframe['rsi_fast'] < 35) &
                    (dataframe['close'] < (
                                dataframe[f'ma_buy_{self.base_nb_candles_buy.value}'] * self.low_offset.value)) &
                    (dataframe['EWO'] > self.ewo_high.value) &
                    (dataframe['rsi'] < self.rsi_buy.value) &
                    (dataframe['volume'] > 0) &
                    (dataframe['close'] < (
                                dataframe[f'ma_sell_{self.base_nb_candles_sell.value}'] * self.high_offset.value))

            )
        )

        conditions.append(
            (

                    (dataframe['rsi_fast'] < 35) &
                    (dataframe['close'] < (
                                dataframe[f'ma_buy_{self.base_nb_candles_buy.value}'] * self.low_offset.value)) &
                    (dataframe['EWO'] < self.ewo_low.value) &
                    (dataframe['volume'] > 0) &
                    (dataframe['close'] < (
                                dataframe[f'ma_sell_{self.base_nb_candles_sell.value}'] * self.high_offset.value))

            )
        )

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x | y, conditions),
                'buy'
            ] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        conditions = []

        conditions.append(
            ((dataframe['close'] > dataframe['hma_50']) &
             (dataframe['close'] > (
                         dataframe[f'ma_sell_{self.base_nb_candles_sell.value}'] * self.high_offset_2.value)) &
             (dataframe['rsi'] > 50) &
             (dataframe['volume'] > 0) &
             (dataframe['rsi_fast'] > dataframe['rsi_slow'])

             )
            |
            (
                    (dataframe['close'] < dataframe['hma_50']) &
                    (dataframe['close'] > (
                                dataframe[f'ma_sell_{self.base_nb_candles_sell.value}'] * self.high_offset.value)) &
                    (dataframe['volume'] > 0) &
                    (dataframe['rsi_fast'] > dataframe['rsi_slow'])
            )

        )

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x | y, conditions),
                'sell'
            ] = 1

        return dataframe

  ```

**Comment by xmatthias** [comment]
While you did test the latest develop - you didn't switch to the [develop documentation](https://www.freqtrade.io/en/latest/advanced-hyperopt/#overriding-pre-defined-spaces) - which has this case fixed since yesterday (10d0987f49b8179f840895882bccebb732098192).

The problem is that the "self" argument is not necessary (actually not allowed) - so the mistake was in the documentation - which for stable, will only be fixed with the next release.

**Comment by EnzovdWetering** [comment]
Thanks for the info @xmatthias! =)
